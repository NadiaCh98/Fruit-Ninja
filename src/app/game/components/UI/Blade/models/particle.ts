export class Particle {

    constructor(private x: number, y: number) {
    }
}
