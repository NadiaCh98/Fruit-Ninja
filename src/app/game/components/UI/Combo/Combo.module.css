.wrapper {
  width: 150px;
  animation: combo 2s 0.7s forwards;
}

.wrapper p {
  font-size: 2rem;
}

.wrapper h2 {
  font-size: var(--text-size);
}

.wrapper p,
h2 {
  background: linear-gradient(
    180deg,
    rgb(255, 255, 255) 0%,
    rgb(212, 212, 212) 74%,
    rgb(180, 180, 180) 100%
  );
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
}

@keyframes combo {
  from {
    transform: scale(1);
  }
  to {
    transform: scale(0);
  }
}
