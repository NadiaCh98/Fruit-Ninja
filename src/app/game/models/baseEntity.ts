export type Id = number | string;

export interface BaseEntity {
    readonly id: Id;
}