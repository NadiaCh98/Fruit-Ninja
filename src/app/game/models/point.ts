export interface Point {
    readonly x: number;
    readonly y: number;
}

export interface Delta extends Point{
    readonly delta: Point;
}