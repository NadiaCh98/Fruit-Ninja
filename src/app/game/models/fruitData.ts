import { Fruits, FruitSpeed } from '../../common/constant';
import { BaseEntity } from './baseEntity';
import { Point } from './point';

export type Bomb = 'bomb';
export type UniformFruit<T extends string> = Lowercase<T>;
export type FruitName = UniformFruit<keyof typeof Fruits>;

export type FruitSile = `${FruitName}_cut`;
export type FruitSlices =  `${FruitSile}1` | `${FruitSile}2`;
export type GeneratableFruit = FruitName | UniformFruit<Bomb>;
export type Fruit = UniformFruit<GeneratableFruit | FruitSlices | Bomb>;

export interface FruitFlyData extends BaseEntity {
    readonly type: GeneratableFruit;
    readonly startPositionX: number;
    readonly flyDirection: Point;
}

export interface FruitData extends FruitFlyData {
    readonly speed: FruitSpeed;
}

export interface FruitSequence {
    readonly delayBetweenFruits: number;
    readonly fruits: FruitData[];
}