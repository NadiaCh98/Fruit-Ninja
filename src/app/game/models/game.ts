import { GeneratableFruit } from './fruitData';
import { GameMode } from '../../common/constant';
import { Point } from './point';

export type Nullable<T> = T | null;
export type NullableNumber = Nullable<number>;

export interface GameConfig {
  readonly game: GameMode;
  readonly timer: Nullable<number>;
  readonly attempts: Nullable<number>;
}

export interface CutFruit {
  readonly fruit: GeneratableFruit;
  readonly point: Point;
}

export interface Combo {
  readonly id: number;
  readonly amount: number;
  readonly point: Point;
}

export type BestScore = Record<GameMode, number>;
