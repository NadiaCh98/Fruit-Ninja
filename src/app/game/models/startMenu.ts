import { GameMode } from './../../common/constant';
import { FruitName } from './fruitData';

export interface StartMenuItem {
  readonly mode: GameMode;
  readonly fruit?: FruitName;
}
