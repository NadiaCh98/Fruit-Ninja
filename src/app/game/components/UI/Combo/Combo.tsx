import React, { useState, useEffect } from 'react';
import { timer } from 'rxjs';
import { tap } from 'rxjs/operators';
import { Combo } from '../../../models/game';
import { GameControls } from '../../GameControls/GameControls';
import styles from './Combo.module.css';

interface ComboProps {
  readonly comboInfo: Combo;
  readonly unmount: (item: Combo) => void;
}

export const COMBO_TIME = 3000;

export const ComboBonus: React.FC<ComboProps> = ({ comboInfo, unmount }) => {
  const {
    amount,
    point: { x: pointX, y: pointY },
  } = comboInfo;
  const [hide, setHide] = useState(false);

  useEffect(() => {
    setHide(false);

    const timeout = timer(COMBO_TIME)
      .pipe(
        tap(() => {
          setHide(true);
          unmount(comboInfo);
        })
      )
      .subscribe();

    return () => timeout.unsubscribe();
  }, [comboInfo, unmount]);

  return !hide ? (
    <GameControls style={{ left: pointX, top: pointY }}>
      <div className={styles.wrapper}>
        <p>{amount} Fruits Combo</p>
        <h2>x 2</h2>
      </div>
    </GameControls>
  ) : null;
};
