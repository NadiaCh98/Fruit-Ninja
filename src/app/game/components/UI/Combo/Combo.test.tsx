import { render, waitFor } from '@testing-library/react';
import React from 'react';
import { Combo } from '../../../models/game';
import { ComboBonus, COMBO_TIME } from './Combo';
describe('Combo', () => {
  const unmount = jest.fn();
  const comboInfo: Combo = {
    id: 0,
    point: {
      x: 100,
      y: 100,
    },
    amount: 4,
  };
  const combo = <ComboBonus comboInfo={comboInfo} unmount={unmount} />;

  it('has a correct template', () => {
    const { container } = render(combo);
    expect(container).toMatchSnapshot();
  });

  it(`asserts that since ${COMBO_TIME}ms container is unmounted and unmount() fire`, async () => {
    const { container } = render(combo);

    await waitFor(
      () => {
        expect(container.innerHTML).toBe('');
        expect(unmount).toBeCalledTimes(1);
        expect(unmount).toBeCalledWith(comboInfo);
      },
      { timeout: 3000 }
    );
  });
});
