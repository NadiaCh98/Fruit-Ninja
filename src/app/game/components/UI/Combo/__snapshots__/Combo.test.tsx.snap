// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combo has a correct template 1`] = `
<div>
  <div
    class="sc-bdnxRM lhbDim"
    style="left: 100px; top: 100px;"
  >
    <div
      class="wrapper"
    >
      <p>
        4
         Fruits Combo
      </p>
      <h2>
        x 2
      </h2>
    </div>
  </div>
</div>
`;
